### ISSUE TITLE
FreqAI - historic predictions are set to 0 when the bot has been restarted and no continuous retraining is configured

### ISSUE THREAD

**Issue by Smidelis** [freqtrade/freqtrade#10801]
## Describe your environment

  * Operating system: Debian GNU/Linux 11 (bullseye)
  * Python Version: 3.12.7 (`python -V`)
  * CCXT version: 4.4.14 (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2024.10-dev-4b582b9d7 (`freqtrade -V` or `docker compose run --rm freqtrade -V` for Freqtrade running in docker)
  
Note: All issues other than enhancement requests will be closed without further comment if the above template is deleted or not filled out.

## Describe the problem:
When running FreqAI and the bot is being stopped (tmux ctrl + c) and started again the historic predictions are 0.

### Steps to reproduce:

  1. Take the default FreqaiExampleStrategy
  2. Take the default config (config_examples/config_freqai.example.json)
  3. Change the value of `"live_retrain_hours"` from the default value 0 to 1 so that it only retrains once per hour.
  4. Run the freqai quickstart command `freqtrade trade --config config_examples/config_freqai.example.json --strategy FreqaiExampleStrategy --freqaimodel LightGBMRegressor --strategy-path freqtrade/templates`
  5. Leave it running for a few hours, so that it retrains a few times.
  6. Restart the bot within a candle so that offline candles are not filled with a 0.
  
### Observed Results:

  * What happened?
  The historic predictions are 0 (targets, di_values, do_predict) from the time on where the last training of the total pairlist has been finished. In the logs there is the message `2024-10-15 15:06:38,472 - freqtrade.freqai.freqai_interface - INFO - Total time spent training pairlist 21.77 seconds`. From that point on, the historical predictions are 0, even though new models have been trained for the pair in the meantime. I have attached two screenshots of frequi and my log files. Screenshot 1 is before the reboot, everything is fine. Screenshot 2 is after the restart with the candle marked from which on the historical predictions are 0. This is the same time as stated in the logs when my entire pair list has finished training.
  * What did you expect to happen?
  The historic predictions not being 0 except the bot has been offline for a candle or more.

### Relevant code exceptions or logs

Find the full log attached and two screenshots of FreqUI with the situation before the restart and after the restart and the corresponding times.
[FreqaiExampleStrategy_historic_predictions_issue.log](https://github.com/user-attachments/files/17381665/FreqaiExampleStrategy_historic_predictions_issue.log)
![Before restart](https://github.com/user-attachments/assets/f03ce0b3-e1eb-4e76-8a67-0121993c66d7)
![After Restart_start zero](https://github.com/user-attachments/assets/7a1fc3e0-817f-4ac5-a297-a0a203b20e4c)

**Comment by robcaulk** [comment]
Thanks @Smidelis for the well investigated bug!

1. You say that you need to change `live_retrain_hours` to something besides 0. So when it is 0, everything works as expected?

**Comment by Smidelis** [comment]
With the default setting from the example config (set to 0) it's not happening. I guess it has something to do with the point in time when the full pairlist has been trained. With the sample strategy and the example config, a training of the full pairlist is happening within a 3m candle.
I am assuming that if you increase the features or something else so that the full pairlist training takes longer than a candle, that the issue will occur too. 
I will test it and give you feedback on it. I will increase the pairs and increase the features, so that the overall training will take quite some time and leave `live_retrain_hours` set to 0.

**Comment by robcaulk** [comment]
Ok @Smidelis, I have reproduced it and located the reason for this. However, I don't know if it is a proper bug (but might deserve improvement depending on what we decide here).

## Bug correction
Just up front - FreqAI will indeed restart making predictions correctly after the reload, so predictions are not 0s for candles that occur *after* the restart (screenshot below. So the bug is actually that you lost some past predictions, not that you cannot make new predictions. This is confirmed with the mechanisms described below.

![image](https://github.com/user-attachments/assets/c1e81cb0-4e88-42e1-84d9-5dba90660aa4)

## FreqAI internal crash resiliency mechanism
So the issue is indeed because you set `live_retrain_hours` to 1 hour. Let me elaborate the internal mechanisms of FreqAI retraining process:

We retrain your model for you periodically, accoridng to the value you set in `live_retrain_hours`. So that is 1 hour in your case.
We are making predictions and collecting those predictions in ram, for performance purposes.
After *a* model is trained, we save historic_predictions.pkl to disk, so that we can reload from a crash.

Now, you decided to retrain once per hour, so your crash resiliency is only guaranteed up to the last time a model was trained, which for your case, could be 1 hour ago. 

When you go to reload FreqAI, we load your historic_predictions.pkl from disk and check when the last saved prediction was. In your case, it was 1 hour ago. Since that is our last know candle, and it is 1 hour in the past from where our current candle stands, we fill in the "lost" candles with 0s so that your predictions in FreqUI stay aligned.

I believe the only thing affected is the visualization in FreqUI (unless you are using historic predictions to generate new predictions, which may be the case for QuickAdapter variants)

## Immediate solution 1
So the solution here is to that if your FreqUI visualization/historic_predictions integrity is of utmost importance, you would want to set `live_retrain_hours` to 0, and then your historic_predictions.pkl on disk will always be as updated as possible (at least as long as it takes to train a single model), and when you go to reload, you will have all your past predictions in FreqUI for your viewing pleasures.

## Proposed solution 2 (probably should get some feedback from @xmatthias on this one)
If you can offer me some rationale for why solution 1 is not suitable, we can add a new parameter that forces the saving of historic_predictions.pkl to disk after every single inference. But I will warn you, this is how we had FreqAI setup originally, and it can become an inference bottleneck as your historic_predictions dataframe grows large (especially if you have huge feature sets). That is why we moved it into the retraining cycle.

## Non-proposed solution 3 (probably should get some commentary from @xmatthias here as well)
The *real* solution would be to convert the collection of historic_predictions into something like duckDB, where we can upsert single rows without needing to resave the entire dataframe. But that is not a proposed solution since:

a) it requires users to run a separate service, which could be added to docker compose, but Freqtrade is aimed at ease of use, so I don't think this is a good idea
b) it would require a general refactor of the code, and I would love to see it, but I dont have the time to see that refactor through from conception, to implementation, to stability guarantees, and to re-documentation. 

## Insert your thoughts below

**Comment by xmatthias** [comment]
I'm not sure if solution1 is a good approach ... as there's influencing things that shouldn't influence each other.

We'd assume to force constant retraining for having uptodate historic_predictions in case of restarts - but that seems like an odd constraint to me.

If i'd like to retrain once per day (`live_retrain_hours` set to 24 - maybe for resource constraint reasons) - why would my inference history only be saved once a day, or i need to retrain constantly (instead of 1 per day).
one is training - one is inference  ...

> If you can offer me some rationale for why solution 1 is not suitable, we can add a new parameter that forces the saving of historic_predictions.pkl to disk after every single inference. But I will warn you, this is how we had FreqAI setup originally, and it can become an inference bottleneck as your historic_predictions dataframe grows large (especially if you have huge feature sets).

I'm not entirely sure on the precise requirement for having this file grow indefinitely (other than for UI niceness) - but i think this could just as well be a rolling version of the last X (say 5000) candles (UI won't allow yo to look back further anyway) - so it'll not grow indefinitely, which would reduce the problematic.

I also see little reason this couldn't be split into "one file per pair" (instead of pickling a dict) ... while the code would become slightly more complex, it'd not cause bottlenecks on bigger pairlists, and would probably also allow for a more efficient fileformat (parquet / feather / ...), which then might support appending, too.


> Non-proposed solution 3 

Don't intend to add another service to run alongside - and if we were to add it - it should be optional (so the complexity in freqtrade would be non-neglectible).

Don't think it's  worth the problems though (we'd need to help users in case of problems with now yet another service we don't control) - especially not for this usecase - considering that this is really a "so the UI looks okish" feature (at least the way i understand `historic_predictions.pkl` for now).

**Comment by Smidelis** [comment]
It's not just a "ui looks okish" feature.
If you do any calculations based on the historical predictions, they will be messed up if there are 0s.
For example using the predicted returns target in your `custom_exit` callback like `current returns > target returns`. If I use this and target returns is 0 after a restart my trades will exit too early. This is because the callbacks are calculated per bot loop and not on a new candle. So by the time the callbacks are processed the prediction may still be 0 which can lead to unwanted behavior.

**Comment by robcaulk** [comment]
@xmatthias 

> I'm not entirely sure on the precise requirement for having this file grow indefinitely

For analytics and reproducibility.

We can add a parameter to let users truncate this file to whatever number of candles they want. Then people who care about analytics can choose how much history they need. Then I suppose we can save the file at the end of a single inference run (so once per candle, all pairs at the same time). But it will still reduce inference performance, so it really isn't recommended.

**Comment by xmatthias** [comment]
>  But it will still reduce inference performance, so it really isn't recommended.

Splitting the file would increase performance again though - most likely into the "won't be really noticed" realm.
I don't really see the benefit  pickling a dict containing dataframes provides over writing individual dataframes - which will result in smaller files, and hence better performance, especially if performant formats (feather, parquet) are used.

**Comment by robcaulk** [comment]
@xmatthias 
> Splitting the file would increase performance again though - most likely into the "won't be really noticed" realm.

Luckily, I made a PR about this already in the past. We did not see the performance gain we hoped for.

https://github.com/freqtrade/freqtrade/pull/8383#issuecomment-1529376877

Now if we did merge that, we need backwards compatibility with anyone running an existing FreqAI strat. I tend to agree with my past self - too intrusive/risky without enough reward.

**Comment by xmatthias** [comment]
don't you need to differentiate a bit more?

current behavior is "move it to training loop for performance writing one large file".
I do understand that moving this to the inference loop won't fly (also because you'd overwrite the file once per pair per loop).
the PR you linked has essentially the same behavior - but with individual files (so instead of writing one large file - you batch-write the same data to 20 files (but all still at the same time) - so as a bunch - you don't gain much - as it's the same process - but writing to 20 files instead of one.

What i'm talking/thinking about  is "move it to inference loop **and** store per pair".
You gain functionality (no downtime if you don't want to retrain constantly) - and you get the performance per pair you need when storing individual files - as you'll not dump all files at once - but only dump them right after inference.

this second part is not what the PR is about (at least not at first glance) - as it still only dumped them all in one batch - though probably most of it could be salvaged - except for `save_historic_predictions_to_disk` probably - which would need a "pair" parameter - and would only store that pair (which allows you to move this from the training loop to right after the inference).

**Comment by robcaulk** [comment]
Hey @xmatthias, it doesn’t work like that. You get all pairs and then
return to the strategy. So whether you do 1 second per pair on 10 pairs or
10 seconds once at the end, it’s the same performance hit on your inference
before you get to your entry.

**Comment by xmatthias** [comment]
> You get all pairs and then return to the strategy

err no ? 
freqAI is called once per pair - and returns that "Per pair" to the strategy.
In my understanding of freqAI (correct me if I'm wrong) - inference in dry/live will happen as part of the `.start()` call - for the pair it's been called.


It's a wrong assumption that the first bot loop after a new candle will always have all pairs. It's most often the case - but not something we guarantee, or aim to guarantee (it would mean delaying all pairs - at the risk of never actually calling the strategy if one pair never comes (for whatever odd reason)).

so sometimes pairs are "late" to refresh (exchange data is late) - so you get 
* loop at :01 - with 5 pairs
* loop at :06 - with the remaining 3 pairs 

So yes, doing 10s once, or doing 10x 1s will be the same - except that it won't - because it may be spread across multiple loops.

---

This feels like a stale discussion though - you don't seem see the benefit in changing this / don't want to change this - so if you see this as intended functioning / no bug, you can also add a hint in the documentation about this as "this is intended - if you rely on having historic predictions always populated, don't use `live_retrain_hours!=0`".

**Comment by robcaulk** [comment]
Ok, yes sorry I meant that you need to populate all pairs indicators before
getting to the entry anyways. But if you say sometimes we populate only
part of the list, for example 7/10, still in this case it is opening saving
and closing 7 files before the entry gets made. So you still have the
penalty on your entry timing since you wanted 7 separate file I/O.

**Comment by Smidelis** [comment]
Just to add a bit more information from tests.
It really is only happening when you set anything else than 0 for `live_retrain_hours`.
Im currently using the example config with the following changes:
* `27 pairs in total`
* `"train_period_days": 90,`
* `"include_shifted_candles": 16,`

As a result a pair takes roughly 115s for training and the total pairlist takes 3455 seconds which is roughly 57 minutes.

I don't get it why in this scenario it's not happening but when you set `live_retrain_hours==0` it is. Isn't a pair getting trained every 57 minutes with `live_retrain_hours==0` basically the same as a pair getting trained every 1h with `live_retrain_hours==1`? In case you are interested i can attach the logs, but i guess there isn't something new in there.

**Comment by robcaulk** [comment]
@Smidelis you said that a pair takes 115s to train. So your historic predictions is saved to disk every 115s.

**Comment by Smidelis** [comment]
Oh, OK, sorry. Now I have it, hopefully.
This means that if I have a highly complex prediction model and whatever I set in `live_retrain_hours`, if the training time of a pair is longer than a candle, the issue will occur?

**Comment by robcaulk** [comment]
It depends on when you reload the bot. If you can align your bot reloading to be right after a pair finishes training.

Actually, I just realized (@xmatthias). We built this protection for you already (https://github.com/freqtrade/freqtrade/blob/develop/freqtrade/freqai/freqai_interface.py#L201). The shutdown procedure waits for training to complete before actually shutting down fully. So if you hit `ctrl-c` one time, we protect you, wait for a train to complete, save your historic predictions, and gracefully shut your bot down. Then you have no issue.

If you click `ctrl-c` `ctrl-c` to cancel the bot in the middle of training operations, then we cannot guarantee historic predictions is properly saved to disk for you.
 
 Keep in mind, if you use `live_retrain_hours: 1`, then the issue can still occur if there is *not* an on-going train happening when you shut your bot down.
